browser/brave_vpn/win/brave_vpn_wireguard_service/status_tray/status_tray_runner.cc

```diff
@@ -164,22 +182,62 @@ void StatusTrayRunner::OnConnected(bool success) {
   VLOG(1) << __func__ << ":" << success;
 }
 
-void StatusTrayRunner::UpdateIconState(bool connected, bool error) {
-  if (!status_tray_ || !status_tray_->GetStatusIcon()) {
+brave_vpn::mojom::ConnectionState StatusTrayRunner::GetConnectionState() {
+  if (IsTunnelServiceRunning()) {
+    return brave_vpn::mojom::ConnectionState::CONNECTED;
+  }
+
+  auto state_from_storage = brave_vpn::wireguard::GetConnectionState();
+  if (state_from_storage.has_value()) {
+    return static_cast<brave_vpn::mojom::ConnectionState>(
+        state_from_storage.value());
+  }
+
+  return brave_vpn::mojom::ConnectionState::DISCONNECTED;
+}
+
+bool StatusTrayRunner::IsIconCreated() {
+  return status_tray_ && status_tray_->GetStatusIcon();
+}
+
+void StatusTrayRunner::UpdateConnectionState() {
+  auto state = GetConnectionState();
+  if (state == brave_vpn::mojom::ConnectionState::CONNECTED) {
+    // Check if we have obsolete connected state in storage.
+    state = IsTunnelServiceRunning()
+                ? state
+                : brave_vpn::mojom::ConnectionState::DISCONNECTED;
+    // if Tunnel service launched it means we have connected state and should
+    // reset storage states because it could be expired from closed browser.
+    wireguard::WriteConnectionState(static_cast<int>(state));
+  }
+
+  if (current_state_.has_value() && current_state_.value() == state) {
```

Comment:
totally optional: I think absl::optional's `operater==` covers has_value() so could be simplified

```suggestion
  if (current_state_ == state) {
```